ydb/query/transaction.py

```diff
@@ -0,0 +1,369 @@
+    return request
+
+
+def _create_rollback_transaction_request(session_state, tx_state):
+    request = _apis.ydb_query.RollbackTransactionRequest()
+    request.tx_id = tx_state.tx_id
+    request.session_id = session_state.session_id
+    return request
+
+
+@base.bad_session_handler
+def wrap_tx_begin_response(rpc_state, response_pb, session_state, tx_state, tx):
+    message = _ydb_query.BeginTransactionResponse.from_proto(response_pb)
+    issues._process_response(message.status)
+    tx_state._change_state(QueryTxStateEnum.BEGINED)
+    tx_state.tx_id = message.tx_meta.tx_id
+    return tx
+
+
+@base.bad_session_handler
+@reset_tx_id_handler
+def wrap_tx_commit_response(rpc_state, response_pb, session_state, tx_state, tx):
+    message = _ydb_query.CommitTransactionResponse.from_proto(response_pb)
+    issues._process_response(message.status)
+    tx_state.tx_id = None
+    tx_state._change_state(QueryTxStateEnum.COMMITTED)
+    return tx
+
+
+@base.bad_session_handler
+@reset_tx_id_handler
+def wrap_tx_rollback_response(rpc_state, response_pb, session_state, tx_state, tx):
+    message = _ydb_query.RollbackTransactionResponse.from_proto(response_pb)
+    issues._process_response(message.status)
+    tx_state.tx_id = None
+    tx_state._change_state(QueryTxStateEnum.ROLLBACKED)
+    return tx
+
+
+class BaseTxContext(base.IQueryTxContext):
+    def __init__(self, driver, session_state, session, tx_mode=None):
+        """
+        An object that provides a simple transaction context manager that allows statements execution
+        in a transaction. You don't have to open transaction explicitly, because context manager encapsulates
+        transaction control logic, and opens new transaction if:
+
+        1) By explicit .begin() and .async_begin() methods;
+        2) On execution of a first statement, which is strictly recommended method, because that avoids useless round trip
+
+        This context manager is not thread-safe, so you should not manipulate on it concurrently.
+
+        :param driver: A driver instance
+        :param session_state: A state of session
+        :param tx_mode: A transaction mode, which is a one from the following choices:
+         1) SerializableReadWrite() which is default mode;
+         2) OnlineReadOnly();
+         3) StaleReadOnly().
+        """
+        self._driver = driver
+        if tx_mode is None:
+            tx_mode = _ydb_query_public.QuerySerializableReadWrite()
+        self._tx_state = QueryTxState(tx_mode)
+        self._session_state = session_state
+        self.session = session
+        self._finished = ""
+        self._prev_stream = None
+
+    def __enter__(self):
+        """
+        Enters a context manager and returns a session
+
+        :return: A session instance
+        """
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        """
+        Closes a transaction context manager and rollbacks transaction if
+        it is not rolled back explicitly
```

Comment:
what about change "not rolled back" to "not finished"? transaction may be committed